apps/aevm/src/aevm_eeevm.erl

```diff
@@ -1424,3 +1408,92 @@ log_topics(AccountAddress, Topics) ->
+%% CALL
+%% ------------------------------------------------------------------------
+
+recursive_call(State, Op) ->
+    CallDepth = aevm_eeevm_state:calldepth(State),
+    case CallDepth < 1024 of
+        false -> {0, State}; %% TODO: Should this consume gas?
+        true  -> recursive_call(CallDepth, State, Op)
+    end.
+
+recursive_call(CallDepth, StateIn, Op) ->
+    %% Message-call into an account.
+    %% i ≡ µm[µs[3] . . .(µs[3] + µs[4] − 1)]
+    State0            = spend_call_gas(StateIn, Op),
+
+    {Gas, State1}     = pop(State0),
+    {To, State2}      = pop(State1),
+    {Value, State3}   = case Op of
+                            ?CALL         -> pop(State2);
+                            ?CALLCODE     -> pop(State2);
+                            ?DELEGATECALL -> {aevm_eeevm_state:value(State2), State2}
+                        end,
+    {IOffset, State4} = pop(State3),
+    {ISize, State5}   = pop(State4),
+    {OOffset, State6} = pop(State5),
+    {OSize, State7}   = pop(State6),
+    Dest              = case Op of
+                            ?CALL -> To;
+                            ?CALLCODE -> aevm_eeevm_state:address(State6);
+                            ?DELEGATECALL -> aevm_eeevm_state:address(State6)
+                        end,
+    {I, State8}       = aevm_eeevm_memory:get_area(IOffset, ISize, State7),
+    GasAfterSpend     = aevm_eeevm_state:gas(State8),
+    Code              = aevm_eeevm_state:extcode(To, State8),
+    CallGas = case Value =/= 0 of
+                  true -> ?GCALLSTIPEND + Gas;
+                  false -> Gas
+              end,
+    case GasAfterSpend >= CallGas of
+        true  -> ok;
+        false -> eval_error(out_of_gas, State7)
+    end,
+    Caller = case Op of
+                 ?CALL -> aevm_eeevm_state:address(State8);
+                 ?CALLCODE -> aevm_eeevm_state:address(State8);
+                 ?DELEGATECALL -> aevm_eeevm_state:caller(State8)
+             end,
+    CallState = aevm_eeevm_state:prepare_for_call(Caller, Dest, CallGas, Value,
+                                                  Code, CallDepth,
+                                                  State8),
+    case aevm_eeevm_state:no_recursion(State8) of
+        true  -> %% TODO: Set callcreates
+            GasOut = GasAfterSpend + CallGas,
+            State9 = aevm_eeevm_state:set_gas(GasOut, State8),
+            State10 = aevm_eeevm_state:add_callcreates(#{ data => I
+                                                        , destination => Dest
+                                                        , gasLimit => CallGas
+                                                        , value => Value
+                                                        }, State9),
+            {1, State10};
+        false ->
+            %% TODO: Should this not count towards mem gas?
+            CallState1 = aevm_eeevm_memory:write_area(0, I, CallState),
```

Comment:
Should we not put the calldata in the `data` field in the state and then have the called contract load it into memory using `CALLDATALOAD`/`CALLDATACOPY`?

Reply:
Yes. (There was a TODO comment for it higher up...)
Fixed now.